nlp/model/augmented_lagrangian.py

```diff
@@ -128,6 +129,21 @@ def hess(self, *args, **kwargs):
         return self.hop(*args, **kwargs)
 
 
+    def magical_step(self, x, g):
+        """Compute a magical step for the slack variables.
+
+        This step minimizes the augmented Lagrangian with respect to the slack
+        variables only for a fixed set of decision variables. Here, g is the 
+        gradient computed by self.grad at x.
+        """
+        model = self.model
+        on = model.original_n
+        m_step = np.zeros(self.n)
+        m_step[on:] = -g[on:] / self.penalty
+        m_step = projected_step(x, m_step, model.Lvar, model.Uvar)
+        return m_step
+
+
```

Comment:
So it seems the current "magical" step is a Newton step (as opposed to a step that actually minimizes the objective). Watch out for the terminology. When we say magical step, we mean an optimal step, which is typically possible when variables are separable (as in the case of slack vars). I never realized the old Auglag didn't do that. I wonder what the impact on performance would be.